File: Week_02/cpp/src/_199_BinaryTreeRightSideView.cpp
#include <vector>
#include <queue>
using namespace std;
struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};
class UsingBfs
{
public:
    vector<int> rightSideView(TreeNode *root)
    {
        vector<int> ret;
        if (!root)
            return ret;
        queue<TreeNode *> q;
        q.push(root);
        while (!q.empty())
        {
            int size = q.size();
            while (size--)
            {
                auto n = q.front();
                q.pop();
                if (n->left)
                    q.push(n->left);
                if (n->right)
                    q.push(n->right);
                if (!size)
                    ret.push_back(n->val);
            }
        }
        return ret;
    }
};
class UsingDfs
{
public:
    vector<int> rightSideView(TreeNode *root)
    {
        vector<int> ret;
        dfs(root, 0, ret);
        return ret;
    }
    void dfs(TreeNode *root, int depth, vector<int> &ret)
    {
        if (!root)
            return;
        if (ret.size() == depth)
            ret.push_back(root->val);
        dfs(root->right, ++depth, ret);
        dfs(root->left, depth, ret);
    }
};

File: Week_02/kotlin/src/_199_BinaryTreeRightSideView.kt
import java.util.*
import kotlin.collections.ArrayList

class _199_Dfs {
    // 根右左
    fun rightSideView(root: TreeNode?): List<Int> {
        val ret = arrayListOf<Int>()
        fun dfs(node: TreeNode?, depth: Int) {
            node ?: return
            if (ret.size == depth) ret += node.`val`
            dfs(node.right, depth + 1)
            dfs(node.left, depth + 1)
        }
        dfs(root, 0)
        return ret
    }

    class Bfs {
        fun rightSideView(root: TreeNode?): List<Int> {
            root ?: return emptyList()
            val q = LinkedList<TreeNode>().apply { offer(root) }
            val ret = ArrayList<Int>()
            while (q.isNotEmpty()) {
                var size = q.size
                while (size-- > 0) {
                    val n = q.poll()
                    n.left?.let { q.offer(it) }
                    n.right?.let { q.offer(it) }
                    if (size == 0) ret += n.`val`
                }
            }
            return ret
        }
    }
}